### ISSUE TITLE
[BUG] Using ParallelFoldFittingStrategy Results in Non-deterministic Out Of Fold Prediction Probabilities and Non-deterministic L2 Models

### ISSUE THREAD

**Issue by LennartPurucker** [autogluon/autogluon#3572]
## Bug 
**Describe the bug**
When using AutoGluon Tabular, the L2 models can be non-deterministic even though they should be deterministic due to using the same input data, random seeds, and being run on CPU. For example, consider the following output (produced by the script below) when running AutoGluon twice:

```bash
Leaderboards:
                     model  score_test  score_val
1    KNeighborsDist_BAG_L1    0.621345   0.653015
6  RandomForestGini_BAG_L1    0.614035   0.669128
7           XGBoost_BAG_L1    0.640351   0.744737
0          CatBoost_BAG_L1    0.694932   0.761404
2          LightGBM_BAG_L1    0.658869   0.762658
3   NeuralNetFastAI_BAG_L1    0.714425   0.775755
4    NeuralNetTorch_BAG_L1    0.701754   0.781957
                   model  score_test  score_val
5  NeuralNetTorch_BAG_L2    0.691033   0.804666
                     model  score_test  score_val
1    KNeighborsDist_BAG_L1    0.621345   0.653015
6  RandomForestGini_BAG_L1    0.614035   0.669128
7           XGBoost_BAG_L1    0.640351   0.744737
0          CatBoost_BAG_L1    0.694932   0.761404
2          LightGBM_BAG_L1    0.658869   0.762658
3   NeuralNetFastAI_BAG_L1    0.714425   0.775755
4    NeuralNetTorch_BAG_L1    0.701754   0.781957
                   model  score_test  score_val
5  NeuralNetTorch_BAG_L2    0.690058   0.805141
```

**Expected behavior**
Running AutoGluon twice with the same data and random seeds should produce the same output when training models on CPU. 

**To Reproduce**
Install current mainline of AutoGluon via source. Then, install `openml` and run the following script. 

<details>
<summary>
Script
</summary>

```python
import numpy as np
import openml
import pandas as pd
from autogluon.tabular import TabularPredictor


def get_data(tid: int, fold: int):
    # Get Task and dataset from OpenML and return split data
    oml_task = openml.tasks.get_task(tid, download_splits=True, download_data=True,
                                     download_qualities=False, download_features_meta_data=False)

    train_ind, test_ind = oml_task.get_train_test_split_indices(fold)
    X, *_ = oml_task.get_dataset().get_data(dataset_format='dataframe')

    return X.iloc[train_ind, :].reset_index(drop=True), X.iloc[test_ind, :].reset_index(drop=True), \
        oml_task.target_name, oml_task.task_type != 'Supervised Classification'


def _print_lb(leaderboard, task_id):
    # print("### Results for ", task_id)
    with pd.option_context("display.max_rows", None, "display.max_columns", None, "display.width", 1000):
        print(leaderboard[leaderboard['model'].str.endswith('L1')].sort_values(by='score_val', ascending=True))
        print(leaderboard[~leaderboard['model'].str.endswith('L1')].sort_values(by='score_val', ascending=False))


def _run(task_id, metric, test=True, fold=0, print_res=True, n_repeat=10):
    l2_train_data, l2_test_data, label, regression = get_data(task_id, fold)

    # Run AutoGluon
    print(f"### task {task_id} and fold {fold} and test {test} shape {(l2_train_data.shape, l2_test_data.shape)}.")
    predictor = TabularPredictor(eval_metric=metric, label=label, verbosity=4)
    predictor.fit(
        train_data=l2_train_data,
        hyperparameters={
            1: {
                'NN_TORCH': {},
                'GBM': [
                    {},
                ],
                'CAT': {},
                'XGB': {},
                'FASTAI': {},
                'RF': [
                    {'criterion': 'gini',
                     'ag_args': {'name_suffix': 'Gini', 'problem_types': ['binary', 'multiclass']}},
                ],
                'KNN': [
                    {'weights': 'distance', 'ag_args': {'name_suffix': 'Dist'}},
                ],
            },
            2: {
                "NN_TORCH": [{}],
            }
        },
        num_stack_levels=1,
        num_bag_sets=n_repeat,
        num_bag_folds=8,
        fit_weighted_ensemble=False,
        num_gpus=0,
        ag_args_ensemble=dict(
           # fold_fitting_strategy="sequential_local",
        ),
    )
    leaderboard = predictor.leaderboard(l2_test_data, silent=True)[['model', 'score_test', 'score_val']].sort_values(
        by='model').reset_index(drop=True)

    if print_res:
        _print_lb(leaderboard, task_id)

    return leaderboard


def random_test():
    print('Fitting first time...')
    leaderboard_a = _run(359955, "roc_auc", fold=2, n_repeat=10)

    print('Fitting second time...')
    leaderboard_b = _run(359955, "roc_auc", fold=2, n_repeat=10)

    print('Leaderboards:')
    _print_lb(leaderboard_a, -1)
    _print_lb(leaderboard_b, -1)

    print('Assert equality')
    pd.testing.assert_frame_equal(leaderboard_a, leaderboard_b)


if __name__ == '__main__':
    random_test()
```

</details>

## Solution

To my understanding, the bug occurs due to the numerical precision of `np.float32` of the running average of the Out Of Fold (OOF) predictions when using `ParallelFoldFittingStrategy`. In essence, depending on the order at which predictions (in `np.float32` format) of fold models are added to the OOF predictions, different numerical rounding errors occur that result in a different vector/matrix of OOF predictions in the end. Furthermore, the order of operations behaves randomly because different fold models might be trained in different order of finish in different order when using the `ParallelFoldFittingStrategy`. 

This difference is also visible at L1 because one can observe (marginally) different scores for L1 models between runs when using a high precision to compute, e.g., log loss. However, this is likely rarely spotted due to rounding nor is this a problem as the L1 models are all trained on the same data and thus are deterministic.  But at L2 the models are now trained on (marginally) different data which can result in visibly different models and results (likely also made more extreme due to #2779). 

To make the L2 models consistent, we need to fix the order of operations when using `ParallelFoldFittingStrategy` or increase the numerical precision of the OOF predictions. I prefer the latter and implemented it in this branch: https://github.com/LennartPurucker/autogluon/tree/conistent_ray_oof. 

Here is the resulting output of the example script:

```bash
Leaderboards:
                     model  score_test  score_val
1    KNeighborsDist_BAG_L1    0.621345   0.653015
6  RandomForestGini_BAG_L1    0.614035   0.669128
7           XGBoost_BAG_L1    0.640351   0.744737
0          CatBoost_BAG_L1    0.694932   0.761404
2          LightGBM_BAG_L1    0.658869   0.762658
3   NeuralNetFastAI_BAG_L1    0.714425   0.775755
4    NeuralNetTorch_BAG_L1    0.701754   0.781957
                   model  score_test  score_val
5  NeuralNetTorch_BAG_L2    0.693957   0.806372
                     model  score_test  score_val
1    KNeighborsDist_BAG_L1    0.621345   0.653015
6  RandomForestGini_BAG_L1    0.614035   0.669128
7           XGBoost_BAG_L1    0.640351   0.744737
0          CatBoost_BAG_L1    0.694932   0.761404
2          LightGBM_BAG_L1    0.658869   0.762658
3   NeuralNetFastAI_BAG_L1    0.714425   0.775755
4    NeuralNetTorch_BAG_L1    0.701754   0.781957
                   model  score_test  score_val
5  NeuralNetTorch_BAG_L2    0.693957   0.806372
```

## Appendix 
<details>
<summary>
Installed Versions
</summary>

```bash
INSTALLED VERSIONS
------------------
date                   : 2023-10-09
time                   : 21:18:15.688474
python                 : 3.8.10.final.0
OS                     : Linux
OS-release             : 5.15.0-1017-aws
Version                : #21~20.04.1-Ubuntu SMP Fri Aug 5 11:44:14 UTC 2022
machine                : x86_64
processor              : x86_64
num_cores              : 128
cpu_ram_mb             : 507863
cuda version           : None
num_gpus               : 0
gpu_ram_mb             : []
avail_disk_size_mb     : 1916480

accelerate             : 0.21.0
autogluon              : 0.8.3b20231009
autogluon.common       : 0.8.3b20231009
autogluon.core         : 0.8.3b20231009
autogluon.eda          : 0.8.3b20231009
autogluon.features     : 0.8.3b20231009
autogluon.multimodal   : 0.8.3b20231009
autogluon.tabular      : 0.8.3b20231009
autogluon.timeseries   : 0.8.3b20231009
boto3                  : 1.28.62
catboost               : 1.2.2
defusedxml             : 0.7.1
evaluate               : 0.4.0
fastai                 : 2.7.12
gluonts                : 0.13.5
hyperopt               : 0.2.7
imodels                : 1.3.18
ipython                : 8.12.3
ipywidgets             : 8.1.1
jinja2                 : 3.1.2
joblib                 : 1.3.2
jsonschema             : 4.17.3
lightgbm               : 3.3.5
lightning              : 2.0.9.post0
matplotlib             : 3.6.3
missingno              : 0.5.2
mlforecast             : 0.9.3
networkx               : 3.1
nlpaug                 : 1.1.11
nltk                   : 3.8.1
nptyping               : 2.4.1
numpy                  : 1.24.4
omegaconf              : 2.2.3
onnxruntime-gpu        : 1.15.1
openmim                : 0.3.9
pandas                 : 2.0.3
phik                   : 0.12.3
Pillow                 : 9.5.0
psutil                 : 5.9.5
pydantic               : 1.10.13
PyMuPDF                : 1.21.1
pyod                   : 1.0.9
pytesseract            : 0.3.10
pytorch-metric-learning: 1.7.3
ray                    : 2.6.3
requests               : 2.28.2
scikit-image           : 0.19.3
scikit-learn           : 1.3.1
scikit-learn-intelex   : None
scipy                  : 1.10.1
seaborn                : 0.12.2
seqeval                : 1.2.2
setuptools             : 60.2.0
shap                   : 0.41.0
skl2onnx               : 1.15.0
statsforecast          : 1.4.0
statsmodels            : 0.14.0
suod                   : 0.0.9
tabpfn                 : 0.1.9
tensorboard            : 2.14.0
text-unidecode         : 1.3
timm                   : 0.9.7
torch                  : 2.0.1
torchmetrics           : 1.0.3
torchvision            : 0.15.2
tqdm                   : 4.65.2
transformers           : 4.31.0
ujson                  : 5.8.0
vowpalwabbit           : 9.8.0
xgboost                : 1.7.6
yellowbrick            : 1.5
```

</details>
